### ISSUE TITLE
[BUG] NiftiLabelsMasker's region_ids_ and region_names_ keys do not match signals array

### ISSUE THREAD

**Issue by tsalo** [nilearn/nilearn#5544]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [ ] Linux
- [x] Mac
- [ ] Windows

### Operating system version

For example one of the following:
- Mac OS Version 15.5 "Sequoia"

### Python version

- [ ] 3.13
- [ ] 3.12
- [ ] 3.11
- [ ] 3.10
- [x] 3.9

### nilearn version

0.12.0, but replicated on 0.12.1.dev31+g8789531e9

### Expected behavior

Based on the NiftiLabelsMasker docstring, I would expect `masker.region_ids_[0]` to equal the value of the first parcel in the atlas image (3).

### Current behavior & error messages

`masker.region_ids_.keys()` and `masker.region_names_.keys()` are both equal to [2, 3, 4] instead of [0, 1, 2]. I think this is because the keys come from the index (not the 'index' column, just the index) of the `lut` after removing rows for parcels that don't exist in the atlas image. I think just resetting the index of the `lut` DataFrame after removing missing parcels would fix the issue, but I could be wrong.

### Steps and code to reproduce bug

```python
import nibabel as nb
import numpy as np
import pandas as pd
from nilearn.maskers import NiftiLabelsMasker

# Create an atlas that is missing the first two regions due to resampling
arr = np.zeros((3, 3, 3))
arr[0, 0, 0] = 3
arr[0, 0, 1] = 4
arr[0, 0, 2] = 5
atlas_img = nb.Nifti1Image(arr.astype(np.int32), np.eye(4))
lut = pd.DataFrame(
    columns=['index', 'name'],
    data=[[1, 'Region A'], [2, 'Region B'], [3, 'Region C'], [4, 'Region D'], [5, 'Region E']],
)

masker = NiftiLabelsMasker(atlas_img, lut=lut)
signals = masker.fit_transform(atlas_img)
# I would expect keys of these dicts to be 0, 1, 2 (index into signals array)
print(masker.region_ids_)
print(masker.region_names_)
print(signals.shape)
```
This outputs:

```
{2: 3, 3: 4, 4: 5}
{2: 'Region C', 3: 'Region D', 4: 'Region E'}
(3,)
```

**Comment by htwangtw** [comment]
Potentially related to #5542 ?

**Comment by tsalo** [comment]
When I read through #5542 I didn't think it was related because your bug arose specifically in the absence of a `lut`/`labels` input, but the underlying problem could be the same.

**Comment by dangom** [comment]
Sorry for jumping in, but I've been hit by this recently as well.

Interestingly this doesn't happen with `labels`:

```python
import nibabel as nb
import numpy as np
import pandas as pd
from nilearn.maskers import NiftiLabelsMasker

arr = np.zeros((3, 3, 3))
arr[0, 0, 0] = 3
arr[0, 0, 1] = 4
arr[0, 0, 2] = 5
atlas_img = nb.Nifti1Image(arr.astype(np.int32), np.eye(4))
labels = ['Region A', 'Region B', 'Region C', 'Region D', 'Region E']

masker = NiftiLabelsMasker(atlas_img, labels=labels)
signals = masker.fit_transform(atlas_img)

print(masker.region_ids_)
print(masker.region_names_)
print(signals.shape)
```

```
sklearn/utils/_set_output.py:316: UserWarning: Too many names for the indices. Dropping excess names values.

print(masker.region_ids_))
{0: 0, 1: 3, 2: 4, 3: 5}

In [152]: print(masker.region_names_))
{1: 'Region A', 2: 'Region B', 3: 'Region C'}

In [153]: print(signals.shape))
(3,)
```

Which made me wonder: To fix these issues, should NiftiLabelsMasker keep internally exclusively a dict of value:label pairs? If passed a DataFrame, it could internally convert it to dict via `dict(zip(lut['index'], lut['name']))`. If passed a list, `dict(zip(range(1, len(labels)+1), labels)`. If passed a dict, there would be no extra preprocessing. This would eliminate the mutually exclusive keywords, make the mapping between regions and IDs cleaner (easier to check for errors and validate) and easier to maintain (would [eliminate this entire path](https://github.com/nilearn/nilearn/blob/8789531e96b88a9e89c84e10b853410218d452a0/nilearn/maskers/nifti_labels_masker.py#L643)).
I searched the issues to see if this had already been discussed but didn't find a related thread.

**Comment by tsalo** [comment]
With `labels`, the region names should be 'Region C', 'Region D', and 'Region E' (corresponding to values 3, 4, 5 in the image), so that seems like a bug as well.

**Comment by dangom** [comment]
Interesting. At the same time, the masker could not have known whether those regions were coded starting from 1.
I think this is exactly the core of the issue: there are heuristics to handle labels, and they may not always match the caller's expectations. 

I'll give a try at a PR over the weekend.

**Comment by Remi-Gau** [comment]
FYI
was planning to work on this when back from holidays on Monday.
And I would definitely prefer to have a pandas dataframe to track regions indices and names whether the user passes labels or an LUT.
Having different implementation for each will probably not make things more maintainable.

**Comment by dangom** [comment]
ok sounds good.
